### PR TITLE
Use Project Lombok to tackle boilerplate code

### PR DESCRIPTION
**Pull request by hnicke**
I'd like to use [Project Lombok](https://projectlombok.org/) for the implementation of the EffectiveModel.
If you don't like this suggestion please give quick feedback. Otherwise I will begin tomorrow to add Lombok to the EffectiveModel implemenation branch (will save me from writing much boilerplate code).

### CODE REVIEW

File: docs/adr/0003-lombok.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+* `+` *the `delombok` feature makes it possible to easily rip lombok out of the project if the need arises*
+* `+` *No boilerplate that can get out of sync if class changes*
+* `-` *introduces more "black magic" to the codebase*
+* `-` *developers have to know what each annotation does*
+* `-` *every developer has to install a plugin for his IDE. supported: intellij, eclipse*
+
+### *Write boilerplate code*
+
+* `+` *pure vanilla java*
+* `-` *boilerplate*
```
Comment (c-mueller) [inline docs/adr/0003-lombok.md]: Once is enough

File: docs/adr/0003-lombok.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+* `-` *more boilerplate*
+* `-` *boilerplate code that might get out of sync when class changes*
+* `-` *did I mention boilerplate*
+
+### *Use kotlin instead of java*
+
+* `+` *lean language design*
+* `+` *would have similar positive effects than Lombok in terms of maintainability and readability*
+* `-` *team doesn't know kotlin, and it's much harder to learn than learning Lombok*
```
Comment (c-mueller) [inline docs/adr/0003-lombok.md]: Well i know Kotlin. :P

File: docs/adr/0003-lombok.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+#### Pro's and cons
+* `+` *significantly smaller EffectiveModel code base*
+* `+` *better maintainability due to smaller code base*
+* `+` *better readability due to smaller code base*
+* `+` *the `delombok` feature makes it possible to easily rip lombok out of the project if the need arises*
+* `+` *No boilerplate that can get out of sync if class changes*
+* `-` *introduces more "black magic" to the codebase*
+* `-` *developers have to know what each annotation does*
+* `-` *every developer has to install a plugin for his IDE. supported: intellij, eclipse*
```
Comment (c-mueller) [inline docs/adr/0003-lombok.md]: This is a huge "problem" (it has to be mentioned in the readme) Please open a issue or update the readme yourself once the plugin is needed

File: docs/adr/0003-lombok.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
+# *Use Project Lombok in order to avoid writing boilerplate code*
+
+**User Story:** 
+
+The implemenation of the EffectiveModel must make use of the builder pattern (many optional parameters required for object construction).
```
Comment (c-mueller) [inline docs/adr/0003-lombok.md]: I think using this is only needed in the model, all other pojo style classes are not large enough to need any features of the lombok thing

**Comment by c-mueller** [comment]
Well i forgot to write this in the Review comment itself:

Writing stupid boilerplate code should be avoided, therefore using this whenever a BuilderPattern is needed is completely okay for me. But i would keep it at that.

File: docs/adr/0003-lombok.md (hunk 3 above)
Comment (nfode) [reply docs/adr/0003-lombok.md]: It has to be added to the Readme but to install a Plugin is not a *huge* problem in my opinion.

File: docs/adr/0003-lombok.md (hunk 1 above)
Comment (nfode) [reply docs/adr/0003-lombok.md]: Seems like he does not like boilerplate code? 😄

File: docs/adr/0003-lombok.md (hunk 2 above)
Comment (nfode) [reply docs/adr/0003-lombok.md]: I have the plan to learn it...

File: docs/adr/0003-lombok.md (hunk 4 above)
Comment (nfode) [reply docs/adr/0003-lombok.md]: Limiting it to the model might be a good idea.

File: docs/adr/0003-lombok.md (hunk 3 above)
Comment (hnicke) [reply docs/adr/0003-lombok.md]: I will add it to the readme when adding the dependency to the project.

**Comment by hnicke** [comment]
@c-mueller you said you would like to limit the use of lombok to the builder pattern.
I'd also limit the usage -- but would also like use these features:
- getters/setters. 
    - reduces even more boilerplate code
    - if a more complicated getter/setter is needed, one can simply provide an implementation in this special case, and lombok will honour that.
    - if javadocs are needed, lombok supports a special notation which converts field javadocs to getter/setter javadocs, which should be sufficient
    - only drawback: theres no annotation for generating Optional<Value> getters. But we can simply write our own getters in these cases
- equals/hash
    - another brilliant option to reduce boilerplate, as writing correct equals/hash methods is tedious
    - i don't like how intellij generates these, and would write them manually if we don't use lombok
    - of course, if a more complex equals method is required, it can be supplied manually, and lombok will honour that
    - especially equals/hash methods tend to get out of sync, which leads to mysterious bugs. not so with lombok.

I'm curious what you think.

**Comment by hnicke** [comment]
## Example: Capability.java

@c-mueller @nfode This is what I would want to do.

#### lombok

```java
package org.opentosca.toscana.model.capability;

import java.util.Set;

import org.opentosca.toscana.model.DescribableEntity;
import org.opentosca.toscana.model.node.RootNode;

import lombok.Builder;
import lombok.Data;
import lombok.NonNull;

/**
 Defines a set of data that can be associated with a {@link RootNode} to describe its capability or feature.
 (TOSCAV1.1, p. 82)
 */
@Builder
@Data
public class Capability extends DescribableEntity {

    /**
     Set of Node Class Types that are valid sources of any relationship established to this capability.
     If empty, all Node classes are valid source types.
     (TOSCAV1.1, p. 83)
     */
    private final Set<Class> validSourceTypes;

    @NonNull
    private final Occurence occurence;
}
```

#### vanilla java
```java
package org.opentosca.toscana.model.capability;

import java.util.Set;

import org.opentosca.toscana.model.DescribableEntity;
import org.opentosca.toscana.model.node.RootNode;

import lombok.NonNull;

/**
 Defines a set of data that can be associated with a {@link RootNode} to describe its capability or feature.
 (TOSCAV1.1, p. 82)
 */
public class Capability extends DescribableEntity {

    /**
     Set of Node Class Types that are valid sources of any relationship established to this capability.
     If empty, all Node classes are valid source types.
     (TOSCAV1.1, p. 83)
     */
    private final Set<Class> validSourceTypes;

    @NonNull
    private final Occurence occurence;

    @java.beans.ConstructorProperties( {"validSourceTypes", "occurence"})
    Capability(Set<Class> validSourceTypes, Occurence occurence) {
        this.validSourceTypes = validSourceTypes;
        this.occurence = occurence;
    }

    public static CapabilityBuilder builder() {
        return new CapabilityBuilder();
    }

    public Set<Class> getValidSourceTypes() {
        return this.validSourceTypes;
    }

    @NonNull
    public Occurence getOccurence() {
        return this.occurence;
    }

    public boolean equals(Object o) {
        if (o == this) return true;
        if (!(o instanceof Capability)) return false;
        final Capability other = (Capability) o;
        if (!other.canEqual((Object) this)) return false;
        final Object this$validSourceTypes = this.getValidSourceTypes();
        final Object other$validSourceTypes = other.getValidSourceTypes();
        if (this$validSourceTypes == null ? other$validSourceTypes != null : !this$validSourceTypes.equals(other$validSourceTypes))
            return false;
        final Object this$occurence = this.getOccurence();
        final Object other$occurence = other.getOccurence();
        if (this$occurence == null ? other$occurence != null : !this$occurence.equals(other$occurence)) return false;
        return true;
    }

    public int hashCode() {
        final int PRIME = 59;
        int result = 1;
        final Object $validSourceTypes = this.getValidSourceTypes();
        result = result * PRIME + ($validSourceTypes == null ? 43 : $validSourceTypes.hashCode());
        final Object $occurence = this.getOccurence();
        result = result * PRIME + ($occurence == null ? 43 : $occurence.hashCode());
        return result;
    }

    protected boolean canEqual(Object other) {
        return other instanceof Capability;
    }

    public String toString() {
        return "Capability(validSourceTypes=" + this.getValidSourceTypes() + ", occurence=" + this.getOccurence() + ")";
    }

    public static class CapabilityBuilder {
        private Set<Class> validSourceTypes;
        private Occurence occurence;

        CapabilityBuilder() {
        }

        public Capability.CapabilityBuilder validSourceTypes(Set<Class> validSourceTypes) {
            this.validSourceTypes = validSourceTypes;
            return this;
        }

        public Capability.CapabilityBuilder occurence(Occurence occurence) {
            this.occurence = occurence;
            return this;
        }

        public Capability build() {
            return new Capability(validSourceTypes, occurence);
        }

        public String toString() {
            return "Capability.CapabilityBuilder(validSourceTypes=" + this.validSourceTypes + ", occurence=" + this.occurence + ")";
        }
    }
}
```

**Comment by nfode** [comment]
@hnicke the shortness of the lombok code speaks for itself

**Pull request merged**